Emulated termination criteria

yallup opened the pull request:
# Description

Aim is to introduce a termination criteria, such that given a set of chains read in to anesthetic we can determine if the nested sampling runs satisfactorily terminated.

This is achieved by introducing a boolean `is_terminated` function, which calls the more informative `critical_ratio`. The aim of this is to return the ratio of _something_ in the live points to the same something contained in the dead points. For first pass this is either the ratio of evidence or the ratio of KL divergences for these two point populations. The former then emulates the PC and MN termination criteria, so the default call of `is_terminated` is the exact polychord setup.


# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code is PEP8 compliant (`flake8 anesthetic tests`)
- [x] My code contains compliant docstrings (`pydocstyle --convention=numpy anesthetic`)
- [x] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have appropriately incremented the [semantic version number](https://semver.org/) in both README.rst and anesthetic/_version.py

yallup committed:
add pc term criteria

yallup commented:
I would suggest black as a pre-commit hook, apologies I tangled some opinionated code formatting up in this, the code style not being enforced as a pre-commit seems vulnerable to such things!

yallup committed:
cover no dead points

yallup committed:
sign change

williamjameshandley committed:
Added read_csv for weighted pandas

williamjameshandley committed:
Added labelled pandas testing

williamjameshandley committed:
Added weighted_labelled_pandas read_csv

williamjameshandley committed:
Added read_csv for NestedSamples

williamjameshandley committed:
Added read_csv to anesthetic

williamjameshandley committed:
Updated pydocstyle

williamjameshandley committed:
bump version to 2.7.1

williamjameshandley committed:
bump version to 2.8.0

williamjameshandley committed:
updated documentation

AdamOrmondroyd commented:
> I would suggest black as a pre-commit hook, apologies I tangled some opinionated code formatting up in this, the code style not being enforced as a pre-commit seems vulnerable to such things!

There is a hook

williamjameshandley committed:
Removed inheritance from documentation

williamjameshandley committed:
Merge branch 'master' into read_csv

williamjameshandley committed:
Merge branch 'master' into read_csv

williamjameshandley committed:
Merge branch 'master' into termination

williamjameshandley committed:
bump version to 2.8.0

williamjameshandley reviewed (changes requested):
In the first instance, you need to revert the black adjustments for now -- the wholesale filechange makes it hard to review. Could you make an issue to discuss code formatters (e.g. switching to black).

williamjameshandley committed:
Merge branch 'master' into read_csv

yallup committed:
remove formatting

yallup committed:
Merge branch 'termination' of github.com:handley-lab/anesthetic into termination

yallup committed:
remove formatting

yallup committed:
hallucinated ref

yallup committed:
Merge branch 'read_csv' into termination

williamjameshandley committed:
Merge branch 'master' into termination

williamjameshandley committed:
Updated to include chain reading

yallup committed:
Merge remote-tracking branch 'origin/read_csv' into termination

yallup committed:
Merge branch 'termination' of github.com:handley-lab/anesthetic into termination

williamjameshandley reviewed (changes requested):
Excellent. Thanks for taking charge of this @yallup -- I've wanted something like this in anesthetic for some time. 

Apologies in advance -- this is going to be quite a picky review. I'd like to make sure that this will generalise to other stopping criteria, and I think the most straightforward way to get there is through a few adjacent iterations.

To make this obvious, we should implement one more stopping criteria. The simplest of these is one suggested to me by John Skilling last year, which is the same as the evidence-based one, but instead operating on the DKL (This is what John Skilling's in-house nested sampling algorithms tend to use). This can be thought of as measuring when the D_KL 'levels off'. John claims anecdotally that this is a more robust criterion.

With this in mind, I suggest a switch of `'Z'` and `'D_KL'` as strings to pick between criteria for now, with `kwarg` being conditionally popped depending on that first switch.

Other stopping criteria we could implement in future are `'logX'`, `'logLmax'`, `'ndead'`.

yallup committed:
Merge branch 'master' into termination

yallup committed:
use inbuilts

yallup committed:
Merge branch 'master' into termination

lukashergt reviewed (commented):
> Introduces a PC termination criteria to check on live points

Would be nice if we could make the description a bit more detailed. Currently this leaves me at a bit of a loss what this is exactly about.

Does this addition deserve an entry in our docs?
https://anesthetic.readthedocs.io/en/latest/

yallup commented on anesthetic/samples.py:
```diff
@@ -1269,6 +1269,43 @@ def truncate(self, logL=None):
+        """
+        logL = self.contour(logL)
+        i_live = ((self.logL >= logL) & (self.logL_birth < logL)).to_numpy()
+        i_dead = ((self.logL < logL)).to_numpy()
```
Not sure I see this, if I am taking the ratio, self.logZ() is the union whereas I want the `self.iloc[dead].logZ()`? So I think I need dead idx? I will implement all the other suggestions and resend for comments

yallup committed:
split out criteria

yallup committed:
Merge branch 'termination' of github.com:handley-lab/anesthetic into termination

yallup commented:
> > Introduces a PC termination criteria to check on live points
> 
> Would be nice if we could make the description a bit more detailed. Currently this leaves me at a bit of a loss what this is exactly about.
> 
> Does this addition deserve an entry in our docs? https://anesthetic.readthedocs.io/en/latest/

Added a more complete description of the goal here to the original pr.

@williamjameshandley Reworked in a slightly unpleasant way to have the two functions declared only in scope, with some currently unnecessary code duplication as it stands, but I wanted to separate them in case the `logX` was added in a different way.

As this currently stand the tests are failing for the D_KL version as the run reports being terminated far too early (if I truncate at an early logL), I suspect my sums may be the wrong way round

williamjameshandley committed:
Suggested refactor

williamjameshandley committed:
Updated documentation

williamjameshandley reviewed (commented):
OK, I've refactored this into something which now gets DKL correctly, and reorganises into something which impacts the samples class a little more gently. I've also implemented a few more criteria which shows why this layout is slightly more general.

The DKL was actually quite subtle, and after some thought it was not obvious how you could easily re-use the calculation from the evidence material.

Updates to documentation/feedback/tests welcome

williamjameshandley committed:
Merge branch 'master' into termination

williamjameshandley committed:
bump version to 2.9.0

williamjameshandley committed:
Updated out-of-date docs

yallup commented:
> OK, I've refactored this into something which now gets DKL correctly, and reorganises into something which impacts the samples class a little more gently. I've also implemented a few more criteria which shows why this layout is slightly more general.
> 
> The DKL was actually quite subtle, and after some thought it was not obvious how you could easily re-use the calculation from the evidence material.
> 
> Updates to documentation/feedback/tests welcome

Thanks Will, much cleaner in namespace when a separate module. My only request, and I am happy to make these changes is to be able to access the value as well as the criteria. As not all are expressed as a ratio my suggestion was to make all functions in `terminate` return a tuple of `(Bool, float)` representing the criteria and the underlying value

If that sounds reasonable I will make those changes to this PR

williamjameshandley commented:
@yallup  it would actually be very useful to have this functionality for @DilyOng's project.

Where did we get to with this?

yallup committed:
Merge

AdamOrmondroyd committed:
fix docs for david

yallup committed:
fix coverage

yallup committed:
Merge branch 'termination' of github.com:handley-lab/anesthetic into termination

AdamOrmondroyd commented:
@yallup see it's happy now

yallup commented:
Should be good to go now, merged in main and fixed coverage @williamjameshandley (thanks @AdamOrmondroyd !)

DilyOng commented:
Thank you very much @yallup for getting this functionality ready!

yallup committed:
correct dict getter

yallup commented on anesthetic/samples.py:
```diff
@@ -1238,6 +1239,26 @@ def truncate(self, logL=None):
+        The remaining arguments are passed to the termination criterion.
+        """
+        try:
+            return termination.__dict__[criterion](self, *args, **kwargs)
```
Good shout, changed

yallup committed:
remove unused line after changing attribute getting

AdamOrmondroyd committed:
Merge remote-tracking branch 'upstream/master' into termination

AdamOrmondroyd reviewed (approved):
lgtm (but please make the summary at the start of the PR accurate)

AdamOrmondroyd commented on anesthetic/termination.py:
```diff
@@ -0,0 +1,41 @@
+def logX(samples, max_logX, nsamples=None):
+    """Terminate when the log-volume of the live points reaches a threshold."""
+    i_live = samples.live_points().index
+    return samples.logX(nsamples)[i_live[0]] < max_logX
```
having now thought about it properly we're all good

yallup commented on anesthetic/termination.py:
```diff
@@ -0,0 +1,41 @@
+import numpy as np
+
+
+def logZ(samples, eps=1e-3, nsamples=None, beta=None):
```
Yes it looks like the indexing isn’t working for nsamples, this should either be fixed or removed. Nsamples here is the same as in other parts of anesthetic stats modules relating to nested sampling. It is a number of replica samples of some estimated quantity to produce. For example if you call samples.logZ(100) you get 100 “samples" of logZ. In this case if it was working for termination you would get a vector of 100 true/false booleans, which would let you potentially get an uncertainty on the termination.

It is very unlikely that this is needed as this is a binary quantity so will be very unlikely to contain any useful information in a set of replicas. I’d probably just remove this as an allowed kwarg. 

This PR does still need approval to be merged, but if you are working on another branch you should be able to merge changes from this branch into another one while this is still a WiP

AdamOrmondroyd committed:
fix for nsamples

AdamOrmondroyd committed:
similar fix for logX

AdamOrmondroyd committed:
sort naming and type, add tests
